### PR TITLE
fix: add missing undefined type to TokensList.links

### PR DESCRIPTION
**Pull request by b6pzeusbc54tvhw5jgpyw8pwz2x6gs**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/markedjs/marked/blob/dfbc9a1efd37e30f0005522ec6755c22d2035fe6/lib/marked.js
- [x] Increase the version number in the header if appropriate.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

### CODE REVIEW

File: types/marked/index.d.ts [hunk 1]
```diff
@@ -119,7 +119,7 @@ declare namespace marked {
 
     type TokensList = Token[] & {
         links: {
-            [key: string]: { href: string; title: string; }
+            [key: string]: { href: string; title: string; } | undefined
```
Comment (BendingBender) [inline types/marked/index.d.ts]: When I look at the sources, I can't see any place where a `link` key is set without a value. This is the location I'm talking about: https://github.com/markedjs/marked/blob/master/lib/marked.js#L441

Is there any other location I'm missing?

**Review by BendingBender — CHANGES_REQUESTED** [review]

**Comment by worr** [comment]
Hmm, I'd like to repro this before approving. I'll repro later today and comment.

El 22 de enero de 2019 8:28:30 a. m. PST, Alfred UC <notifications@github.com> escribió:
>b6pzeusbc54tvhw5jgpyw8pwz2x6gs commented on this pull request.
>
>
>
>> @@ -119,7 +119,7 @@ declare namespace marked {
> 
>     type TokensList = Token[] & {
>         links: {
>-            [key: string]: { href: string; title: string; }
>+            [key: string]: { href: string; title: string; } |
>undefined
>
>There is no case where the key is set without value, but the developer
>can access it with an arbitrary key and the value at this time is
>undefined.
>
>```js
>const someLinkUrl =  tokenArray.links.someKey.href
>```
>In the above example,
>tokenArray.links.someKey may be undefined, but typescript does not
>cause any errors even when use `strictNullChecks: true`.
>
>if `undefined` type is added, `[ts] Object is possibly 'undefined'.
>[2532]` error is occur and developer can change the code safely like
>below:
>
>```diff
>-const someLinkUrl =  tokenArray.links.someKey.href
>+const emptyLink = { href: "", title: "" }
>+const someLinkUrl =  (tokenArray.links.someKey || emptyLink).href
>```
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/DefinitelyTyped/DefinitelyTyped/pull/32350#discussion_r249858184

-- 
Enviado desde mi dispositivo Android con K-9 Mail. Por favor, disculpa mi brevedad.

File: types/marked/index.d.ts (hunk 1 above)
Comment (BendingBender) [reply types/marked/index.d.ts]: I understand what you mean but it is kind of alien to me to add `undefined` type to index signatures.

If you have an array, for example, you know that the indexes you access may be `undefined` at any time, I'm not sure that I've already seen an instance where someone added `undefined` to an array type to have an explicit check for non-existing indexes. Same goes here. It's just convenience of usage, I suppose.

Comment (worr) [reply types/marked/index.d.ts]: I agree with @BendingBender here as well. While I'm sympathetic to the issue here, I don't think that adding this is the right way to fix it given that it invalidates any object's type definition.

Comment (b6pzeusbc54tvhw5jgpyw8pwz2x6gs) [reply types/marked/index.d.ts]: @BendingBender @worr 
Thank you for your explanation. I am a Typescript beginner and I think I misunderstood this part. In the array example below, I have checked there is no error, and I have learned that it is a normal operation.

```typescript
const user: Array<{ name: string }> = []
const userName = user[4].name
```

**Pull request closed without merging**